components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs

```diff
@@ -40,6 +40,67 @@ public class SubscriptionServices : ISubscriptionManager, IDisposable {
             _sessionManager = sessionManager ?? throw new ArgumentNullException(nameof(sessionManager));
             _codec = codec ?? throw new ArgumentNullException(nameof(codec));
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _timer = new Timer(OnCheckAsync, null, kIdleCheckTimespan, Timeout.InfiniteTimeSpan);
+            _errorSignaled = false; 
+        }
+
+        /// <summary>
+        /// notify the owner about exceptions in subscriptions
+        /// </summary>
+        public void SignalSubscriptionError() {
+            if (_errorSignaled != true) {
+                _errorSignaled = true;
+                Try.Op(() => _timer.Change(kFastRetryTimespan, Timeout.InfiniteTimeSpan));
+            }
+        }
+
+        /// <summary>
+        /// Check connectivity
+        /// </summary>
+        private async void OnCheckAsync(object sender) {
+            var success = true;
+            try {
+                Try.Op(() => _timer.Change(Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan));
```

Comment:
This is an example re above, a) fully try catch log (even though log could theoratically throw), b) disable firing of timer so this task can first run and then re-arm the timer for the next task. no action at this point.

Reply:
thanks this is really helpful!